src/app/delegation_backend/src/itn_uptime_analyzer/sheets.go

```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		log.Fatalf("Unable to retrieve data from sheet: %v", err)
+	}
+
+	identityString := strings.Join([]string{identity["public-key"], identity["public-ip"]}, "-")
+
+	for index, row := range resp.Values {
+		if row[0] == identityString {
+			rowIndex = index + 1
+			exactMatch = true
+			break
+		}
+		firstEmptyRow = firstEmptyRow + 1
+	}
+
+	if exactMatch == false {
+		for index, row := range resp.Values {
+			str := fmt.Sprintf("%v", row[0])
+			if strings.Split(str, "-")[0] == identity["public-key"] {
+				rowIndex = index + 1
+				exactMatch = false
+			}
+		}
+	}
+
+	return exactMatch, rowIndex, firstEmptyRow
+}
+
+// Appends the identity string of the node to the first column
+
+func (identity Identity) AppendNext(client *sheets.Service, log *logging.ZapEventLogger) {
+	col := IDENTITY_COLUMN
+	readRange := ITN_UPTIME_ANALYZER_SHEET + "!" + col + ":" + col
+	spId := OutputSpreadsheetId()
+
+	identityString := strings.Join([]string{identity["public-key"], identity["public-ip"]}, "-")
+
+	cellValue := []interface{}{identityString}
+
+	valueRange := sheets.ValueRange{
+		Values: [][]interface{}{cellValue},
+	}
+
+	_, err := client.Spreadsheets.Values.Append(spId, readRange, &valueRange).ValueInputOption("USER_ENTERED").Do()
+	if err != nil {
+		log.Fatalf("Unable to append data to sheet: %v", err)
+	}
+}
+
+// Inserts the identity string of the node in the first column under rowIndex
+
+func (identity Identity) InsertBelow(client *sheets.Service, log *logging.ZapEventLogger, rowIndex int) {
+	col := IDENTITY_COLUMN
+	readRange := fmt.Sprintf("%s!%s%d:%s%d", ITN_UPTIME_ANALYZER_SHEET, col, rowIndex + 1, col, rowIndex + 1)
+	spId := OutputSpreadsheetId()
+
+	identityString := strings.Join([]string{identity["public-key"], identity["public-ip"]}, "-")
+
+	cellValue := []interface{}{identityString}
+
+	valueRange := sheets.ValueRange{
+		Values: [][]interface{}{cellValue},
+	}
+
+	_, err := client.Spreadsheets.Values.Append(spId, readRange, &valueRange).ValueInputOption("USER_ENTERED").InsertDataOption("INSERT_ROWS").Do()
+	if err != nil {
+		log.Fatalf("Unable to insert data in sheet: %v", err)
+	}
+}
+
+// Finds the first empty column on the row specified and puts up or not up
+
+func (identity Identity) AppendUptime(client *sheets.Service, log *logging.ZapEventLogger, rowIndex int) {
+	readRange := fmt.Sprintf("%s!A%d:Z%d", ITN_UPTIME_ANALYZER_SHEET, 1, 1)
+	spId := OutputSpreadsheetId()
+
+	resp, err := client.Spreadsheets.Values.Get(spId, readRange).Do()
+	if err != nil {
+		log.Fatalf("Unable to retrieve data from sheet: %v\n", err)
+	}
+
+	var nextEmptyColumn int = len(resp.Values[0])
+
+	updateRange := fmt.Sprintf("%s!%s%d", ITN_UPTIME_ANALYZER_SHEET, string(nextEmptyColumn + 65), rowIndex)
+
+	var cellValue []interface{}
+
+	uptimeArrayLength, err := strconv.Atoi(identity["uptime"])
+	if err != nil {
+		log.Fatalf("Unable to retrieve data from sheet: %v\n", err)
+	}
+
+	if (uptimeArrayLength >= 47) {
```

Comment:
Let's have a constant for period of sync `const syncPeriod time.Duration` (currently set to 15 minutes) and calculate `expectedIntervals = floor(timeSinceLastSync/syncPeriod)`

Reply:
Motivation: we may change the sync period in future